### DISCUSSION TITLE
GetStaticProps type in conjunction with InferGetStaticPropsType type

### DISCUSSION THREAD

**Discussion by GlennyDC** [vercel/next.js#17600] [Help] [answered]
Hello

It seems not possible to use the `GetStaticProps` type in conjunction with the `InferGetStaticPropsType` type. Do you need to explicitly use the `GetStaticPropsContext` and `GetStaticPropsResult` types to work in conjunction with the `InferGetStaticPropsType` type? Why can't both work together?

The following code correctly types the props of the page component `Blog`, but not the function `getStaticProps`:

```js
import { InferGetStaticPropsType } from 'next'

type Post = {
  author: string
  content: string
}

export const getStaticProps = async () => {
  const res = await fetch('https://.../posts')
  const posts: Post[] = await res.json()

  return {
    props: {
      posts,
    },
  }
}

function Blog({ posts }: InferGetStaticPropsType<typeof getStaticProps>) {
  // will resolve posts to type Post[]
}

export default Blog
```

The following code correctly types the function `getStaticProps` , but not the props of the page component `Blog`:

```js
import { InferGetStaticPropsType, GetStaticProps } from 'next'

type Post = {
  author: string
  content: string
}

export const getStaticProps: GetStaticProps = async () => {
  const res = await fetch('https://.../posts')
  const posts: Post[] = await res.json()

  return {
    props: {
      posts,
    },
  }
}

function Blog({ posts }: InferGetStaticPropsType<typeof getStaticProps>) {
  // will resolve posts to type any
}

export default Blog
```

My dependencies: 

```
"dependencies": {
  "next": "9.5.3",
  "react": "16.13.1",
  "react-dom": "16.13.1"
},
"devDependencies": {
  "@types/node": "^14.11.2",
  "@types/react": "^16.9.49",
  "@types/react-dom": "^16.9.8",
  "typescript": "^4.0.3"
}
```

Kind regards

Glenny

**Comment by lfades** [comment, marked as answer]
> It seems not possible to use the GetStaticProps type in conjunction with the InferGetStaticPropsType type

@GlennyDC You are right! It's no possible to use them together.

This is hard to explain but it goes like this:

If you have the following code:

```tsx
export const getStaticProps = async () => {
  // ...
}
```

TypeScript will infer the type of `getStaticProps`, if you're using VS Code and move your cursor over it you should be able to see the type that TypeScript created for it, that inferred type will also have inferred props, and those are the props `InferGetStaticPropsType` is using.

Now, in the following code:

```tsx
export const getStaticProps: GetStaticProps = async () => {
  // ...
}
```

TypeScript doesn't have to infer the type of `getStaticProps` because it already knows it is `GetStaticProps`, so it won't infer the returned props either, and if you then use `InferGetStaticPropsType` what you get is `{ [key: string]: any }`, which is the value of `P` in the definition of [`GetStaticProps`](https://github.com/vercel/next.js/blob/canary/packages/next/types/index.d.ts#L92-L95), unless you define the props manually:

```ts
type Props = {
 // ...
}

export const getStaticProps: GetStaticProps<Props>
```

But if that's the case, then you don't need `InferGetStaticPropsType` at all 😄

**Comment by robertpiosik** [comment]
Don't forget you can obtain argument type in `getStaticProps` with `GetStaticPropsContext`.

```ts
import { InferGetStaticPropsType, GetStaticPropsContext} from 'next'

type Post = {
  author: string
  content: string
}

export const getStaticProps = async (ctx: GetStaticPropsContext) => {
  const res = await fetch('https://.../posts')
  const posts: Post[] = await res.json()

  return {
    props: {
      posts,
    },
  }
}

function Blog({ posts }: InferGetStaticPropsType<typeof getStaticProps>) {
  // will resolve posts to type Post
}

export default Blog
```
